storage/innobase/handler/ha_innodb.cc

```diff
@@ -6556,10 +6556,6 @@ ha_innobase::open(const char* name, int, uint)
 		}
 	}
 
-	if (table && m_prebuilt->table) {
-		ut_ad(table->versioned() == m_prebuilt->table->versioned());
-	}
```

Comment:
Why it is valid to fail this condition?

Reply:
It's actually
```c++
ut_ad(table->versioned(VERS_TIMESTAMP) == m_prebuilt->table->versioned());
```
which is obviously incorrect. Prebuilt table is created here without (trx_id based) system versioning. And assertion fails.

Instead of deleting this code we may apply this patch and tests will pass:
```diff
diff --git a/storage/innobase/handler/ha_innodb.cc b/storage/innobase/handler/ha_innodb.cc
index 5489b8541eb..06fb75e6260 100644
--- a/storage/innobase/handler/ha_innodb.cc
+++ b/storage/innobase/handler/ha_innodb.cc
@@ -6556,6 +6556,11 @@ ha_innobase::open(const char* name, int, uint)
                }
        }
 
+       if (table && m_prebuilt->table) {
+               ut_ad((table->versioned(VERS_TIMESTAMP) || table->versioned(VERS_TRX_ID))
+                     == m_prebuilt->table->versioned());
+       }
+
        info(HA_STATUS_NO_LOCK | HA_STATUS_VARIABLE | HA_STATUS_CONST | HA_STATUS_OPEN);
        DBUG_RETURN(0);
 }
diff --git a/storage/innobase/handler/handler0alter.cc b/storage/innobase/handler/handler0alter.cc
index 0aef727b434..c2315f15786 100644
--- a/storage/innobase/handler/handler0alter.cc
+++ b/storage/innobase/handler/handler0alter.cc
@@ -4995,7 +4995,8 @@ prepare_inplace_alter_table_dict(
                                field_type |= DATA_UNSIGNED;
                        }
 
-                       if (altered_table->versioned(VERS_TRX_ID)) {
+                       if (altered_table->versioned(VERS_TIMESTAMP)
+                           || altered_table->versioned(VERS_TRX_ID)) {
                                if (i == altered_table->s->row_start_field) {
                                        field_type |= DATA_VERS_START;
                                } else if (i ==
```

It don't know what's better. What do you think?

Reply:
Make `if (altered_table->versioned())` in `handler0alter.cc:4998`. Revert change of `ha_innodb.cc`.

Reply:
Done.